src/data_source/sql.rs

```diff
@@ -0,0 +1,987 @@
+// General Public License for more details.
+// You should have received a copy of the GNU General Public License along with this program. If not,
+// see <https://www.gnu.org/licenses/>.
+
+#![cfg(feature = "sql-data-source")]
+
+use super::VersionedDataSource;
+use crate::{
+    availability::{
+        AvailabilityDataSource, BlockId, BlockQueryData, LeafId, LeafQueryData, QueryResult,
+        TransactionHash, TransactionIndex, UpdateAvailabilityData,
+    },
+    status::{MempoolQueryData, StatusDataSource, UpdateStatusData},
+    Block, Deltas, QueryableBlock, Resolvable,
+};
+use async_std::{net::ToSocketAddrs, task::spawn};
+use async_trait::async_trait;
+use futures::{
+    channel::oneshot,
+    future::{select, Either, FutureExt},
+    stream::BoxStream,
+    task::{Context, Poll},
+    AsyncRead, AsyncWrite,
+};
+use hotshot_types::traits::{
+    metrics::Metrics,
+    node_implementation::{NodeImplementation, NodeType},
+    signature_key::EncodedPublicKey,
+};
+use itertools::Itertools;
+use std::{ops::RangeBounds, pin::Pin};
+use tokio_postgres::{types::BorrowToSql, Client, NoTls, RowStream, ToStatement};
+
+pub use crate::include_migrations;
+pub use anyhow::Error;
+pub use refinery::Migration;
+pub use tokio_postgres::{self as postgres};
+
+// This needs to be reexported so that we can reference it by absolute path relative to this crate
+// in the expansion of `include_migrations`, even when `include_migrations` is invoked from another
+// crate which doesn't have `include_dir` as a dependency.
+pub use include_dir::include_dir;
+
+/// Embed migrations from the given directory into the current binary.
+///
+/// The macro invocation `include_migrations!(path)` evaluates to an expression of type `impl
+/// Iterator<Item = Migration>`. Each migration must be a text file which is an immediate child of
+/// `path`, and there must be no non-migration files in `path`. The migration files must have names
+/// of the form `V${version}__${name}.sql`, where `version` is a positive integer indicating how the
+/// migration is to be ordered relative to other migrations, and `name` is a descriptive name for
+/// the migration.
+///
+/// `path` should be an absolute path. It is possible to give a path relative to the root of the
+/// invoking crate by using environment variable expansions and the `CARGO_MANIFEST_DIR` environment
+/// variable.
+///
+/// As an example, this is the invocation used to load the default migrations from the
+/// `hotshot-query-service` crate. The migrations are located in a directory called `migrations` at
+/// the root of the crate.
+///
+/// ```
+/// # use hotshot_query_service::data_source::sql::{include_migrations, Migration};
+/// let migrations: Vec<Migration> = include_migrations!("$CARGO_MANIFEST_DIR/migrations").collect();
+/// assert_eq!(migrations[0].version(), 10);
+/// assert_eq!(migrations[0].name(), "init_schema");
+/// ```
+#[macro_export]
+macro_rules! include_migrations {
```

Comment:
In case you know, how is this different from https://docs.rs/refinery/0.8.11/refinery/macro.embed_migrations.html ?

Reply:
`embed_migrations` gives you a [`Runner`](https://docs.rs/refinery/latest/refinery/struct.Runner.html#), but I just wanted a `Vec<Migration>` so I could add other migrations and modify the order and whatnot (see `validate_migrations`). We will use this capability in the sequencer to insert additional migrations to create indices for sequencer-specific queries.